### ISSUE TITLE
SingletonExtensionFactory will leak PluginClassLoader

### ISSUE THREAD

**Issue by RationalityFrontline** [pf4j/pf4j#487]
SingletonExtensionFactory uses a WeakHashMap as cache, however this is a wrong usage of WeakHashMap, because the weak reference is on the key, not the value, since the value hold strong reference to its PluginClassLoader，the keys will never get garbage collected, and so will the values.

References:
https://eric-liu-developer.medium.com/when-weakhashmap-leaks-d5eec906eeb
https://stackoverflow.com/a/1802879

**Comment by andotorg** [comment]
fix this question， clear in weakhashmap？

**Comment by RationalityFrontline** [comment]
> fix this question， clear in weakhashmap？

@andotorg If do this you will need to decide when to remove the key, then you have to listen to plugin's state, but SingletonExtensionFactory doesn't have access to its PluginManager, so PluginManager need to be pass in as SingletonExtensionFactory's constructor parameter. By the way, WeakHashMap can be replaced with normal HashMap. Below is the solution:
```java
public class SingletonExtensionFactory extends DefaultExtensionFactory {

    private final List<String> extensionClassNames;

    private final Map<ClassLoader, Map<String, Object>> cache;

    public SingletonExtensionFactory(PluginManager pluginManager, String... extensionClassNames) {
        this.extensionClassNames = Arrays.asList(extensionClassNames);

        cache = new HashMap<>();
        
        pluginManager.addPluginStateListener(event -> {
            if (event.getPluginState() != PluginState.STARTED) {
                cache.remove(event.getPlugin().getPluginClassLoader());
            }
        });
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T> T create(Class<T> extensionClass) {
        String extensionClassName = extensionClass.getName();
        ClassLoader extensionClassLoader = extensionClass.getClassLoader();

        if (!cache.containsKey(extensionClassLoader)) {
            cache.put(extensionClassLoader, new HashMap<>());
        }

        Map<String, Object> classLoaderBucket = cache.get(extensionClassLoader);

        if (classLoaderBucket.containsKey(extensionClassName)) {
            return (T) classLoaderBucket.get(extensionClassName);
        }

        T extension = super.create(extensionClass);
        if (extensionClassNames.isEmpty() || extensionClassNames.contains(extensionClassName)) {
            classLoaderBucket.put(extensionClassName, extension);
        }

        return extension;
    }

}
```

**Comment by RationalityFrontline** [comment]
Solved in PR #490
